src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php

```diff
@@ -67,16 +68,29 @@ private function registerHandlers(ContainerBuilder $container)
 
         foreach ($container->findTaggedServiceIds($this->handlerTag, true) as $serviceId => $tags) {
             foreach ($tags as $tag) {
-                $handles = $tag['handles'] ?? $this->guessHandledClass($container, $serviceId);
-
-                if (!class_exists($handles)) {
-                    $messageClassLocation = isset($tag['handles']) ? 'declared in your tag attribute "handles"' : 'declared in `__invoke` function';
-
-                    throw new RuntimeException(sprintf('The message class "%s" %s of service "%s" does not exist.', $messageClassLocation, $handles, $serviceId));
+                $handles = $tag['handles'] ?? $this->guessHandledClasses($r = $container->getReflectionClass($container->getParameterBag()->resolveValue($container->getDefinition($serviceId)->getClass())), $serviceId);
+                if (empty($handles)) {
+                    throw new RuntimeException(sprintf('The handler service "%s" does not seem to handle any message.', $serviceId));
                 }
 
                 $priority = $tag['priority'] ?? 0;
-                $handlersByMessage[$handles][$priority][] = new Reference($serviceId);
+
+                foreach ($handles as $messageClass) {
+                    if (is_array($messageClass)) {
+                        $messagePriority = $messageClass[1];
+                        $messageClass = $messageClass[0];
+                    } else {
+                        $messagePriority = $priority;
+                    }
+
+                    if (!class_exists($messageClass)) {
+                        $messageClassLocation = isset($tag['handles']) ? 'declared in your tag attribute "handles"' : sprintf('used as argument type in method "%s::__invoke()"', $r->getName());
```

Comment:
Might not be worth trying to hack together a fix, but the second part of this message about `__invoke()` is now not necessarily the case... with the new HandlerInterface

Reply:
I think it's a good idea.